### PR TITLE
Added support for jupyter attachment field, embedded image rendering in markdown notebook cells

### PR DESCRIPTION
**Pull request by Yoyokrazy**
Relevant to [#5321](https://github.com/microsoft/vscode-jupyter/issues/5321)

When opening a notebook previously created in jupyter notebooks, images are stored via attachment fields. This PR adds support for the renderer to see this metadata and attachment field, and renders images using an extended markdown-it rule for images.

`extensions/ipynb/yarn.lock` was updated due to needing markdown-it inside of the extensions/ipynb directory

### CODE REVIEW

File: extensions/ipynb/package.json [hunk 1]
```diff
@@ -51,6 +51,16 @@
 				"priority": "default"
 			}
 		],
+    "notebookRenderer": [
+      {
+        "id": "vscode.markdown-it-cell-attachment-renderer-extension",
```
Comment (rebornix) [inline extensions/ipynb/package.json:56]: Nit, this is the id for the renderer so maybe call it `vscode.markdown-it-cell-attachment-renderer`

File: src/vs/workbench/contrib/notebook/browser/view/renderers/webviewMessages.ts [hunk 2]
```diff
@@ -329,6 +330,7 @@ export interface IMarkupCellInitialization {
 	content: string;
 	offset: number;
 	visible: boolean;
+	metadata: NotebookCellMetadata;
```
Comment (rebornix) [inline src/vs/workbench/contrib/notebook/browser/view/renderers/webviewMessages.ts:333]: We might need this for `export interface IShowMarkupCellMessage {` too for markdown update.

**Review by rebornix — COMMENTED** [review]
Looks great in overall!

File: extensions/ipynb/package.json [hunk 3]
```diff
@@ -70,16 +80,19 @@
...
-		"compile": "npx gulp compile-extension:ipynb",
-		"watch": "npx gulp watch-extension:ipynb"
+		"compile": "npx gulp compile-extension:ipynb && npm run build-notebook",
+		"watch": "npx gulp watch-extension:ipynb",
+		"build-notebook": "node ./esbuild"
 	},
 	"dependencies": {
 		"@enonic/fnv-plus": "^1.3.0",
 		"detect-indent": "^6.0.0",
+    "markdown-it": "^12.3.2",
```
Comment (mjbvz) [inline extensions/ipynb/package.json:90]: You likely don't need this dependency as you are only using the types from `markdown-it`

File: extensions/ipynb/src/cellAttachmentRenderer.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+	if (!markdownItRenderer) {
+		throw new Error(`Could not load 'vscode.markdown-it-renderer'`);
+	}
+
+	markdownItRenderer.extendMarkdownIt((md: MarkdownIt) => {
+		const original = md.renderer.rules.image;
+		md.renderer.rules.image = (tokens: MarkdownItToken[], idx: number, options, env, self) => {
+			const token = tokens[idx];
+			const src = token.attrGet('src');
+			const attachments: Record<string, Record<string, string>> = (env.outputItem.metadata as any).custom.attachments;
```
Comment (mjbvz) [inline extensions/ipynb/src/cellAttachmentRenderer.ts:25]: You may need to check that `env.outputItem.metadata` and `env.outputItem.metadata.custom` are not undefined before accessing properties on them

**Review by mjbvz — CHANGES_REQUESTED** [review]

File: src/vs/workbench/contrib/notebook/browser/view/renderers/webviewMessages.ts (hunk 2 above)
Comment (Yoyokrazy) [reply src/vs/workbench/contrib/notebook/browser/view/renderers/webviewMessages.ts:333]: Not totally sure why `IShowMarkupCellMessage` would need a field for metadata, but I can add it there as well?

File: extensions/ipynb/src/cellAttachmentRenderer.ts (hunk 4 above)
Comment (Yoyokrazy) [reply extensions/ipynb/src/cellAttachmentRenderer.ts]: Newest commit should address this, I think that 
`attachments = env.outputItem.metadata?.custom?.attachments` + `if (attachments && src)` 
should handle it?

**Review by mjbvz — APPROVED** [review]

**Comment by alexdima** [comment]
@mjbvz @rebornix When running `yarn` on `main`, there are changes to `extensions/ipynb/yarn.lock`, could you please take a look.

**Comment by rebornix** [comment]
@alexdima thanks I'll take a look and fix

**Pull request merged**